### ISSUE TITLE
Extract keys from a config?

### ISSUE THREAD

**Issue by jhenahan** [dhall-lang/dhall-haskell#423]
Is it possible (or even meaningful) to extract the keys from a Dhall file?

I'm using Dhall for a system validation tool, and one part of that is a fairly long list of `sysctl` settings. I have a Dhall file which describes the correct settings, and what I'd like to say is "for all the keys defined in the config file, run `sysctl -q <key>` and capture its value", which I then compare against the record I construct from importing the correct config. I'm using Turtle for the `sysctl` calls, and I'd be happier if the set of keys to search could be defined in a single place (the Dhall file), rather than having to repeat it in the application source.

Is this possible? If not, is there another solution that I haven't considered that gets me something similarly extensible?

**Comment by ocharles** [comment]
It's certainly possible, but you'll have to drop down to the parse tree. I suggest parsing the file, resolving imports, type checking, normalizing, and then finding the `RecordLit` in the resulting `Expr`. This `RecordLit` is basically just a map of keys to expressions. I'm assuming by "extract keys from a config" you mean "extract all field names from a record".

**Comment by jhenahan** [comment]
Yes, that's precisely what I mean. I'll put some work into that and either PR it here or publish a little helper library.

**Comment by ocharles** [comment]
I think it might be a bit too specific to be added to Dhall, but I'll let
Gabriel weigh in on that.

**Comment by Gabriel439** [comment]
Another way to do this is that since you have the record in Haskell, you can use Haskell generic programming support to extract field names.  See an example of how to do this here:

https://stackoverflow.com/a/27818445/1026598

**Comment by Gabriel439** [comment]
I also think there is value in something that performs all the steps to get the abstract syntax tree without marshaling it into a Haskell type so I'd accept a pull request for that

**Comment by sellout** [comment]
I added an `importExprWith` function on my branch for #418, that does this (so the TH fully normalizes rather than just resolves), but I didn’t push it yet. That’s because I wanted to call it from `importWith`, and that got a little messy when dealing with adding the `expected` type annotation for the expression.

I can either get that done on #418, or do it separately.

**Comment by jhenahan** [comment]
I've explored some of the `Generic` machinery, and I can almost get what I want:

I can take a config that looks like

```
{ vm =
    { swappiness = 0 }
, kernel =
    { pid_max =
        4000000
    , shmmax =
        2199023255552
    , msgmni =
        512
    , msgmnb =
        65536
    , sem =
        "250 256000 32 2048"
    }
, fs =
    { filemax = 1000000000, aio_max_nr = 524288 }
, net =
    { ipv4 = { tcp_keepalive_time = 600, tcp_keepalive_intvl = 60 } }
}
```
consume that to construct a record like

```haskell
data SysctlSetting = SysctlSetting { vm :: VmSetting
                                   , kernel :: KernelSetting
                                   , fs :: FsSetting
                                   , net :: NetSetting
                                   } deriving (Show, Eq, Ord, Generic, Data, Typeable)

data VmSetting = VmSetting { swappiness :: Natural
                           } deriving (Show, Eq, Ord, Generic, Data, Typeable)

data KernelSetting = KernelSetting { pid_max :: Natural
                                   , shmmax :: Natural
                                   , msgmni :: Natural
                                   , msgmnb :: Natural
                                   , sem :: Text
                                   } deriving (Show, Eq, Ord, Generic, Data, Typeable)

data FsSetting = FsSetting { filemax :: Natural
                           , aio_max_nr :: Natural
                           } deriving (Eq, Ord, Generic, Data, Typeable)

data NetSetting = NetSetting { ipv4 :: IPv4Setting } deriving (Show, Eq, Ord, Generic, Data, Typeable)

data IPv4Setting = IPv4Setting { tcp_keepalive_time :: Natural
                               , tcp_keepalive_intvl :: Natural
                               } deriving (Show, Eq, Ord, Generic, Data, Typeable)
```

And then I can construct a `HashMap` like

```haskell
settingMap :: SysctlSetting -> HashMap String [String]
settingMap config = fromList $ zip sysSections sysKeys
  where sysKeys = map constrFields . gmapQ toConstr $ config
        sysSections = concatMap constrFields . dataTypeConstrs . dataTypeOf $ (undefined :: SysctlSetting)
```

This gets me really close, except for three things:

1. It feels pretty scary compared to "decompose the config file".
2. I'll have to do some extra work to get into the nested fields in the `net.*` bit.
3. Nested records aren't exactly ergonomic when it comes to extending the thing.

I originally had a flat representation like

```haskell
data SysctlSetting = Sysctl { _key :: Text
                            , _value :: SysctlValue
                            } deriving (Show, Eq, Ord, Generic, Data, Typeable)

data SysctlValue = SCString Text | SCInt Int deriving (Show, Eq, Ord, Generic, Data, Typeable)
```
I like the latter representation better just because it's easier to construct values in the poorly-behaved space of `sysctl` properties, but I like the additional safety (and reduction in repetition in the config file) of the nested record approach.

My eventual goal is to have this tool be consumed by non-programmers at work, so if I can synthesize these two paths such that changes in the config can be made without a corresponding change in the code, I think I'll be really pleased.

I'll have a look at `importExprWith`, while I'm at it.

**Comment by sellout** [comment]
@jhenahan I have just this minute pushed `inputExprWith` to #418 (in case you already looked and couldn’t find it).

**Comment by Gabriel439** [comment]
Like @sellout mentioned, I think this is fixed by the change in #418.  That adds new `Dhall.inputExpr`/`Dhall.inputExprWith` utilities that you can use to read in the raw abstract syntax tree for an expression, and then you can write code that traverses the `Expr` tree to retrieve the keys that you are interested in

**Comment by jhenahan** [comment]
Awesome! I'll have a look at that.
